File: packages/lhc-py/tests/fixtures/model_call.py
"""Ported from packages/lhc/test/fixtures/model-call.ts. Phase 1.

Constants and pure assignment/response maps are REAL. Call builders that
return ModelCall functions are skeletons.
"""

from __future__ import annotations

import asyncio
import copy
from typing import Literal, NotRequired, TypedDict

from lhc.shared_tech.inference_types import (
    ModelAssignment,
    ModelCall,
    ModelCallInput,
    ModelCallOk,
    ModelCallResult,
    ThinkingLevel,
)
from lhc.shared_tech.prompts import DEFAULT_PROMPT_NAMES

DERIVATION_TYPES = (
    "smoothed_prompt",
    "tool_result_summary",
    "turn_rendering",
    "pre_detailed_assembly",
    "detailed_turn_compression",
    "chunk_summary_detailed",
    "chunk_summary_brief",
)
DerivationType = Literal[
    "smoothed_prompt",
    "tool_result_summary",
    "turn_rendering",
    "pre_detailed_assembly",
    "detailed_turn_compression",
    "chunk_summary_detailed",
    "chunk_summary_brief",
]

INFERENCE_DERIVATION_TYPES = (
    "smoothed_prompt",
    "tool_result_summary",
    "detailed_turn_compression",
    "chunk_summary_brief",
)
InferenceDerivationType = Literal[
    "smoothed_prompt",
    "tool_result_summary",
    "detailed_turn_compression",
    "chunk_summary_brief",
]

# The prompt template a fixture assignment selects for a kind: the inference
# default when one exists, else the deterministic type's own template.
def _fixture_prompt(kind: InferenceDerivationType) -> str:
    inference_default = DEFAULT_PROMPT_NAMES.get(kind)
    if inference_default is not None:
        return inference_default
    return "unknown-prompt"


# validAssignments gives every kind a distinct fake provider/model lane;
# recordingCall infers the kind back from the model string, so per-kind
# routing is observable from the log alone (TC-1.2).
FAKE_PROVIDER_PREFIX = "prov-"
FAKE_MODEL_PREFIX = "model-"


class ModelAssignmentOverride(TypedDict, total=False):
    """Partial<ModelAssignment> — TS camelCase data keys."""

    provider: str
    model: str
    prompt: str
    targetMinRatio: float
    targetMaxRatio: float
    targetAimRatio: float
    thinking: ThinkingLevel


def valid_assignments(
    overrides: dict[InferenceDerivationType, ModelAssignmentOverride] | None = None,
) -> dict[InferenceDerivationType, ModelAssignment]:
    """Every kind gets a distinct fake provider/model lane — pure data.

    Applies every override field (provider/model/prompt/ratios/thinking),
    matching TS object spread onto the base assignment.
    """
    overrides = overrides or {}
    result: dict[InferenceDerivationType, ModelAssignment] = {}
    for kind in INFERENCE_DERIVATION_TYPES:
        ov = overrides.get(kind, {})
        result[kind] = ModelAssignment(
            provider=ov["provider"] if "provider" in ov else f"{FAKE_PROVIDER_PREFIX}{kind}",
            model=ov["model"] if "model" in ov else f"{FAKE_MODEL_PREFIX}{kind}",
            prompt=ov["prompt"] if "prompt" in ov else _fixture_prompt(kind),
            target_min_ratio=ov.get("targetMinRatio"),
            target_max_ratio=ov.get("targetMaxRatio"),
            target_aim_ratio=ov.get("targetAimRatio"),
            thinking=ov.get("thinking"),
        )
    return result


# One distinct canned sentence per kind, so cross-kind bleed is visible in
# landed form content, never just in call counts.
def canned_responses() -> dict[DerivationType, str]:
    return {kind: f"canned {kind} text from the fake host" for kind in DERIVATION_TYPES}


class RecordingCallBundle(TypedDict):
    call: ModelCall
    log: list[ModelCallInput]


# Resolves each call with its kind's canned text — the kind inferred from
# the model routing key validAssignments made unique per kind. Every input
# is cloned into `log` in call order. An unknown model is a test bug and
# throws loudly.
def recording_call(responses: dict[DerivationType, str]) -> RecordingCallBundle:
    log: list[ModelCallInput] = []
    known = set(INFERENCE_DERIVATION_TYPES)

    async def call(input: ModelCallInput) -> ModelCallResult:
        log.append(copy.deepcopy(input))
        kind = (
            input.model[len(FAKE_MODEL_PREFIX) :]
            if input.model.startswith(FAKE_MODEL_PREFIX)
            else input.model
        )
        if kind not in known:
            raise RuntimeError(f'recordingCall: no canned response for model "{input.model}"')
        return ModelCallOk(text=responses[kind])  # type: ignore[index]

    return {"call": call, "log": log}


# Returns script entries in order; a call past the end of the script is a
# test bug and throws loudly.
def scripted_call(script: list[ModelCallResult]) -> ModelCall:
    next_index = 0

    async def call(input: ModelCallInput) -> ModelCallResult:
        nonlocal next_index
        if next_index >= len(script):
            raise RuntimeError(
                f'scriptedCall: script exhausted after {len(script)} calls (model "{input.model}")'
            )
        entry = script[next_index]
        next_index += 1
        return entry

    return call


# A host whose function throws — the AC-3.3 containment leg.
def throwing_call(error: Exception) -> ModelCall:
    async def call(_input: ModelCallInput) -> ModelCallResult:
        raise error

    return call


# A host that never settles — the DD-6 timeout leg (tests pass a small
# timeoutMs).
def hanging_call() -> ModelCall:
    async def call(_input: ModelCallInput) -> ModelCallResult:
        await asyncio.Future()  # never settles
        raise AssertionError("unreachable")

    return call

File: packages/lhc-rs/tests/fixtures/model_call.rs
//! Ported from packages/lhc/test/fixtures/model-call.ts.
//!
//! Constants and pure assignment/response maps are REAL. Call builders are
//! boundary doubles (host ModelCall functions) — REAL, matching lhc-py.

use std::collections::HashMap;
use std::ops::{Index, IndexMut};
use std::sync::atomic::{AtomicUsize, Ordering};
use std::sync::{Arc, Mutex};

use indexmap::IndexMap;
use lhc::shared_tech::inference_types::{
    ModelAssignment, ModelCall, ModelCallInput, ModelCallResult, ThinkingLevel,
};
use lhc::shared_tech::prompts::DEFAULT_PROMPT_NAMES;

/// Closed fixture vocabulary for derivation types (TS `DerivationType` /
/// `DERIVATION_TYPES` in model-call.ts). Production keeps plain strings.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum DerivationType {
    SmoothedPrompt,
    ToolResultSummary,
    TurnRendering,
    PreDetailedAssembly,
    DetailedTurnCompression,
    ChunkSummaryDetailed,
    ChunkSummaryBrief,
}

impl DerivationType {
    pub const ALL: &[DerivationType] = &[
        Self::SmoothedPrompt,
        Self::ToolResultSummary,
        Self::TurnRendering,
        Self::PreDetailedAssembly,
        Self::DetailedTurnCompression,
        Self::ChunkSummaryDetailed,
        Self::ChunkSummaryBrief,
    ];

    pub const fn as_str(self) -> &'static str {
        match self {
            Self::SmoothedPrompt => "smoothed_prompt",
            Self::ToolResultSummary => "tool_result_summary",
            Self::TurnRendering => "turn_rendering",
            Self::PreDetailedAssembly => "pre_detailed_assembly",
            Self::DetailedTurnCompression => "detailed_turn_compression",
            Self::ChunkSummaryDetailed => "chunk_summary_detailed",
            Self::ChunkSummaryBrief => "chunk_summary_brief",
        }
    }
}

pub const DERIVATION_TYPES: &[&str] = &[
    DerivationType::SmoothedPrompt.as_str(),
    DerivationType::ToolResultSummary.as_str(),
    DerivationType::TurnRendering.as_str(),
    DerivationType::PreDetailedAssembly.as_str(),
    DerivationType::DetailedTurnCompression.as_str(),
    DerivationType::ChunkSummaryDetailed.as_str(),
    DerivationType::ChunkSummaryBrief.as_str(),
];

/// Closed fixture vocabulary for inference derivation types that cross the
/// host ModelCall boundary (TS `InferenceDerivationType`).
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum InferenceDerivationType {
    SmoothedPrompt,
    ToolResultSummary,
    DetailedTurnCompression,
    ChunkSummaryBrief,
}

impl InferenceDerivationType {
    pub const ALL: &[InferenceDerivationType] = &[
        Self::SmoothedPrompt,
        Self::ToolResultSummary,
        Self::DetailedTurnCompression,
        Self::ChunkSummaryBrief,
    ];

    pub const fn as_str(self) -> &'static str {
        match self {
            Self::SmoothedPrompt => "smoothed_prompt",
            Self::ToolResultSummary => "tool_result_summary",
            Self::DetailedTurnCompression => "detailed_turn_compression",
            Self::ChunkSummaryBrief => "chunk_summary_brief",
        }
    }

    pub fn from_wire(s: &str) -> Option<Self> {
        match s {
            "smoothed_prompt" => Some(Self::SmoothedPrompt),
            "tool_result_summary" => Some(Self::ToolResultSummary),
            "detailed_turn_compression" => Some(Self::DetailedTurnCompression),
            "chunk_summary_brief" => Some(Self::ChunkSummaryBrief),
            _ => None,
        }
    }
}

/// String vocabulary derived from [`InferenceDerivationType`].
pub const INFERENCE_DERIVATION_TYPES: &[&str] = &[
    InferenceDerivationType::SmoothedPrompt.as_str(),
    InferenceDerivationType::ToolResultSummary.as_str(),
    InferenceDerivationType::DetailedTurnCompression.as_str(),
    InferenceDerivationType::ChunkSummaryBrief.as_str(),
];

pub const FAKE_PROVIDER_PREFIX: &str = "prov-";
pub const FAKE_MODEL_PREFIX: &str = "model-";

fn fixture_prompt(kind: InferenceDerivationType) -> String {
    let kind_str = kind.as_str();
    for (k, name) in DEFAULT_PROMPT_NAMES {
        if *k == kind_str {
            return (*name).to_string();
        }
    }
    "unknown-prompt".to_string()
}

/// Partial override bag for [`valid_assignments`] (TS `Partial<ModelAssignment>`).
#[derive(Debug, Clone, Default)]
pub struct ModelAssignmentOverride {
    pub provider: Option<String>,
    pub model: Option<String>,
    pub prompt: Option<String>,
    pub target_min_ratio: Option<f64>,
    pub target_max_ratio: Option<f64>,
    pub target_aim_ratio: Option<f64>,
    pub thinking: Option<ThinkingLevel>,
}

/// Total closed assignment container for the four inference derivation kinds
/// (TS `Record<InferenceDerivationType, ModelAssignment>`).
#[derive(Debug, Clone)]
pub struct InferenceAssignments {
    pub smoothed_prompt: ModelAssignment,
    pub tool_result_summary: ModelAssignment,
    pub detailed_turn_compression: ModelAssignment,
    pub chunk_summary_brief: ModelAssignment,
}

impl InferenceAssignments {
    pub fn get(&self, kind: InferenceDerivationType) -> &ModelAssignment {
        match kind {
            InferenceDerivationType::SmoothedPrompt => &self.smoothed_prompt,
            InferenceDerivationType::ToolResultSummary => &self.tool_result_summary,
            InferenceDerivationType::DetailedTurnCompression => &self.detailed_turn_compression,
            InferenceDerivationType::ChunkSummaryBrief => &self.chunk_summary_brief,
        }
    }

    pub fn get_mut(&mut self, kind: InferenceDerivationType) -> &mut ModelAssignment {
        match kind {
            InferenceDerivationType::SmoothedPrompt => &mut self.smoothed_prompt,
            InferenceDerivationType::ToolResultSummary => &mut self.tool_result_summary,
            InferenceDerivationType::DetailedTurnCompression => &mut self.detailed_turn_compression,
            InferenceDerivationType::ChunkSummaryBrief => &mut self.chunk_summary_brief,
        }
    }

    /// Explicit conversion for the production [`InferenceConfig`] string-key path.
    pub fn to_string_keyed(&self) -> IndexMap<String, ModelAssignment> {
        let mut map = IndexMap::new();
        for kind in InferenceDerivationType::ALL {
            map.insert(kind.as_str().to_string(), self.get(*kind).clone());
        }
        map
    }
}

impl Index<InferenceDerivationType> for InferenceAssignments {
    type Output = ModelAssignment;

    fn index(&self, kind: InferenceDerivationType) -> &ModelAssignment {
        self.get(kind)
    }
}

impl IndexMut<InferenceDerivationType> for InferenceAssignments {
    fn index_mut(&mut self, kind: InferenceDerivationType) -> &mut ModelAssignment {
        self.get_mut(kind)
    }
}

fn assignment_for(
    kind: InferenceDerivationType,
    ov: Option<&ModelAssignmentOverride>,
) -> ModelAssignment {
    let kind_str = kind.as_str();
    ModelAssignment {
        provider: ov
            .and_then(|o| o.provider.clone())
            .unwrap_or_else(|| format!("{FAKE_PROVIDER_PREFIX}{kind_str}")),
        model: ov
            .and_then(|o| o.model.clone())
            .unwrap_or_else(|| format!("{FAKE_MODEL_PREFIX}{kind_str}")),
        prompt: ov
            .and_then(|o| o.prompt.clone())
            .unwrap_or_else(|| fixture_prompt(kind)),
        target_min_ratio: ov.and_then(|o| o.target_min_ratio),
        target_max_ratio: ov.and_then(|o| o.target_max_ratio),
        target_aim_ratio: ov.and_then(|o| o.target_aim_ratio),
        thinking: ov.and_then(|o| o.thinking),
    }
}

/// Every kind gets a distinct fake provider/model lane — pure data. REAL.
/// Returns a total closed [`InferenceAssignments`] covering all four inference kinds.
pub fn valid_assignments(
    overrides: Option<&HashMap<InferenceDerivationType, ModelAssignmentOverride>>,
) -> InferenceAssignments {
    let empty = HashMap::new();
    let overrides = overrides.unwrap_or(&empty);
    InferenceAssignments {
        smoothed_prompt: assignment_for(
            InferenceDerivationType::SmoothedPrompt,
            overrides.get(&InferenceDerivationType::SmoothedPrompt),
        ),
        tool_result_summary: assignment_for(
            InferenceDerivationType::ToolResultSummary,
            overrides.get(&InferenceDerivationType::ToolResultSummary),
        ),
        detailed_turn_compression: assignment_for(
            InferenceDerivationType::DetailedTurnCompression,
            overrides.get(&InferenceDerivationType::DetailedTurnCompression),
        ),
        chunk_summary_brief: assignment_for(
            InferenceDerivationType::ChunkSummaryBrief,
            overrides.get(&InferenceDerivationType::ChunkSummaryBrief),
        ),
    }
}

/// One distinct canned sentence per kind. REAL.
pub fn canned_responses() -> HashMap<String, String> {
    DERIVATION_TYPES
        .iter()
        .map(|kind| {
            (
                (*kind).to_string(),
                format!("canned {kind} text from the fake host"),
            )
        })
        .collect()
}

pub struct RecordingCallBundle {
    pub call: ModelCall,
    pub log: Arc<Mutex<Vec<ModelCallInput>>>,
}

/// Resolves each call with its kind's canned text — kind inferred from the
/// model routing key [`valid_assignments`] made unique per kind. REAL.
pub fn recording_call(responses: &HashMap<String, String>) -> RecordingCallBundle {
    let log = Arc::new(Mutex::new(Vec::new()));
    let responses = responses.clone();
    let log_for_call = Arc::clone(&log);
    let known: Vec<String> = INFERENCE_DERIVATION_TYPES
        .iter()
        .map(|s| (*s).to_string())
        .collect();
    let call: ModelCall = Box::new(move |input| {
        let responses = responses.clone();
        let log = Arc::clone(&log_for_call);
        let known = known.clone();
        Box::pin(async move {
            log.lock().expect("recordingCall log").push(input.clone());
            let kind = if input.model.starts_with(FAKE_MODEL_PREFIX) {
                input.model[FAKE_MODEL_PREFIX.len()..].to_string()
            } else {
                input.model.clone()
            };
            if !known.iter().any(|k| k == &kind) {
                panic!(
                    "recordingCall: no canned response for model \"{}\"",
                    input.model
                );
            }
            let text = responses.get(&kind).cloned().unwrap_or_else(|| {
                panic!(
                    "recordingCall: no canned response for model \"{}\"",
                    input.model
                );
            });
            ModelCallResult::Ok { text }
        })
    });
    RecordingCallBundle { call, log }
}

/// Returns script entries in order; a call past the end of the script is a
/// test bug and panics loudly. REAL.
pub fn scripted_call(script: Vec<ModelCallResult>) -> ModelCall {
    let next = Arc::new(AtomicUsize::new(0));
    let script = Arc::new(script);
    Box::new(move |input| {
        let next = Arc::clone(&next);
        let script = Arc::clone(&script);
        Box::pin(async move {
            let i = next.fetch_add(1, Ordering::SeqCst);
            match script.get(i) {
                Some(entry) => entry.clone(),
                None => panic!(
                    "scriptedCall: script exhausted after {} calls (model \"{}\")",
                    script.len(),
                    input.model
                ),
            }
        })
    })
}

/// A host whose function panics — the AC-3.3 containment leg (JS Promise.reject).
/// REAL. `safe_call` will catch panics once implemented.
pub fn throwing_call(error: Box<dyn std::error::Error + Send + Sync>) -> ModelCall {
    let msg = error.to_string();
    Box::new(move |_input| {
        let msg = msg.clone();
        Box::pin(async move {
            panic!("{msg}");
        })
    })
}

/// A host that never settles — the DD-6 timeout leg. REAL.
pub fn hanging_call() -> ModelCall {
    Box::new(|_input| Box::pin(futures::future::pending::<ModelCallResult>()))
}
